query/worker.go

```diff
@@ -0,0 +1,194 @@
+
+// Index returns the queryTask's index within the work queue.
+func (q *queryTask) Index() int {
+	return q.index
+}
+
+// jobResult is the final result of the worker's handling of the queryTask.
+type jobResult struct {
+	task *queryTask
+	err  error
+}
+
+// worker is responsible for polling work from the work queue, and hand it to
+// the associated peer. It validates incoming responses with the current
+// query's response handler, and poll more work for the peer when it has
+// successfully received a response to the request.
+type worker struct {
+	// peer is the peer this worker will supply with queries, and handle
+	// responses from.
+	peer Peer
+
+	// nextJob is a channel of queries to be distributed, where the worker
+	// will poll new work tasks from.
+	nextJob <-chan *queryTask
+
+	// results is a channel where the worker will deliver the final result
+	// of the last polled task.
+	results chan<- *jobResult
+
+	quit <-chan struct{}
+}
+
+// run starts the worker. The method is blocking, and should be started in a
+// goroutine.
+func (w *worker) run() {
+	// Subscribe to messages from the peer.
+	msgChan, cancel := w.peer.SubscribeRecvMsg()
+	defer cancel()
+
+	for {
+		var job *queryTask
+
+		select {
+		// Poll a new job from the nextJob channel.
+		case job = <-w.nextJob:
+
+		// Ignore any message received while not working on a task.
+		case <-msgChan:
+			continue
+
+		// If the peer disconnected, we can exit immediately, as we
+		// weren't working on a query.
+		case <-w.peer.OnDisconnect():
+			log.Debugf("Peer %v for worker disconnected",
+				w.peer.Addr())
+			return
+
+		case <-w.quit:
+			return
+		}
+
+		// We received a query job, send it to the peer.
+		w.peer.QueueMessageWithEncoding(
+			job.Req, nil, job.options.encoding,
+		)
+
+		// Wait for the correct response to be received from the peer,
+		// or an error happening.
+		var (
+			resp    wire.Message
+			jobErr  error
+			timeout = time.After(job.options.timeout)
+		)
+
+	Loop:
+		for {
+
+			select {
+			// A message was received from the peer, use the
+			// response handler to check whether it was answering
+			// our request.
+			case resp = <-msgChan:
+				done, progress := job.HandleResp(
+					job.Req, resp, w.peer.Addr(),
+				)
+
+				// If the response did not answer our query, we
+				// check whether it did progress it.
+				if !done {
+					// If it did make progress we reset the
+					// timeout. This ensures that the
+					// queries with multiple responses
+					// expected won't timeout before all
+					// responses have been handled.
+					// TODO(halseth): separete progress
+					// timeout value.
+					if progress {
+						timeout = time.After(
+							job.options.timeout,
+						)
+					}
+					continue Loop
+				}
+
+				// We did get a valid response, and can break
+				// the loop.
+				break Loop
+
+			// If the timeout is reached before a valid response
+			// has been received, we exit with an error.
+			case <-timeout:
```

Comment:
Do we want to prevent reassigning the failed query to the same peer here or we do just rely on it eventually being assigned to a different peer?

Reply:
Idea is that the peer will be deprioritised for next attempt